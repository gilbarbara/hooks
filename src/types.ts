import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
