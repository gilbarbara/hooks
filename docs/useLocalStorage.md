# useLocalStorage

Interact with the browser localStorage API.

## Usage

```tsx
import { useLocalStorage } from '@gilbarbara/hooks';

function Component() {
  const [value, setValue, remove] = useLocalStorage('my-key', 'foo');

  return (
    <div>
      <div>Value: {value}</div>
      <button onClick={() => setValue('bar')}>bar</button>
      <button onClick={() => setValue('baz')}>baz</button>
      <button onClick={() => remove()}>Remove</button>
    </div>
  );
}
```

## Reference

```typescript
type UseLocalStorageOptions<TValue> =
  | {
  raw: true;
}
  | {
  deserializer: (value: string) => TValue;
  raw: false;
  serializer: (value: TValue) => string;
};

type UseLocalStorageResult<TValue> = [
    TValue | undefined,
  Dispatch<SetStateAction<TValue | undefined>>,
  () => void,
];

useLocalStorage<TValue>(key, initialValue?: TValue, options?: UseLocalStorageOptions<TValue>): UseLocalStorageResult;
```

- key — localStorage key to manage.
- initialValue — the initial value to set, if the value in localStorage is empty.
- options.raw — if set to true, the hook will not attempt to JSON serialize stored values.
- options.serializer — custom serializer (defaults to JSON.stringify)
- options.deserializer — custom deserializer (defaults to JSON.parse)
