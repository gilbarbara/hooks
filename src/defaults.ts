import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
