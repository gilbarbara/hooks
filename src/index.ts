export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export { useResponsive } from './useResponsive';
export { useScript } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useUnmount } from './useUnmount';
export { useUpdateEffect } from './useUpdateEffect';
export { useWhyDidYouUpdate } from './useWhyDidYouUpdate';
export { useWindowSize, type WindowSize } from './useWindowSize';
