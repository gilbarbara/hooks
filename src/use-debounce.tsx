import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
