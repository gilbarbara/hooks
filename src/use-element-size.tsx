import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
