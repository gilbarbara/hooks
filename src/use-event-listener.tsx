import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
