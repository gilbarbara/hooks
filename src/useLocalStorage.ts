import {
  Dispatch,
  SetStateAction,
  useCallback,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import { canUseDOM, noop } from './utils';

export type UseLocalStorageOptions<TValue> =
  | {
      raw: true;
    }
  | {
      deserializer: (value: string) => TValue;
      raw: false;
      serializer: (value: TValue) => string;
    };

export type UseLocalStorageResult<TValue> = [
  TValue | undefined,
  Dispatch<SetStateAction<TValue | undefined>>,
  () => void,
];

function useLocalStorageHook<TValue>(
  key: string,
  initialValue?: TValue,
  options?: UseLocalStorageOptions<TValue>,
): UseLocalStorageResult<TValue> {
  if (!key) {
    throw new Error('useLocalStorage: "key" is required');
  }

  const deserializer = useMemo(
    () => (options?.raw ? (value: any) => value : options?.deserializer ?? JSON.parse),
    [options],
  );
  const serializer = useMemo(
    () => (options?.raw ? String : options?.serializer ?? JSON.stringify),
    [options],
  );

  const initializer = useRef((k: string) => {
    try {
      const localStorageValue = localStorage.getItem(k);

      if (localStorageValue !== null) {
        return deserializer(localStorageValue);
      }

      initialValue && localStorage.setItem(k, serializer(initialValue));

      return initialValue;
      /* c8 ignore next 6 */
    } catch {
      // If user is in private mode or has storage restriction
      // localStorage can throw. JSON.parse and JSON.stringify
      // can throw, too.
      return initialValue;
    }
  });

  const [state, setState] = useState<TValue | undefined>(() => initializer.current(key));

  useLayoutEffect(() => setState(initializer.current(key)), [key]);

  const set: Dispatch<SetStateAction<TValue | undefined>> = useCallback(
    patch => {
      try {
        const newState = patch instanceof Function ? patch(state) : patch;

        if (typeof newState === 'undefined') {
          return;
        }

        let value: string;

        if (options) {
          if (options.raw) {
            value = typeof newState === 'string' ? newState : JSON.stringify(newState);
          } else if (options?.serializer) {
            value = options.serializer(newState);
          } else {
            value = JSON.stringify(newState);
          }
        } else {
          value = JSON.stringify(newState);
        }

        localStorage.setItem(key, value);
        setState(deserializer(value));
        /* c8 ignore next 4 */
      } catch {
        // If user is in private mode or has storage restriction
        // localStorage can throw. Also, JSON.stringify can throw.
      }
    },
    [deserializer, key, options, state],
  );

  const remove = useCallback(() => {
    try {
      localStorage.removeItem(key);
      setState(undefined);
      /* c8 ignore next 4 */
    } catch {
      // If user is in private mode or has storage restriction
      // localStorage can throw.
    }
  }, [key, setState]);

  return [state, set, remove];
}

function useLocalStorageSSR<TValue>(
  _key: string,
  initialValue?: TValue,
  _options?: UseLocalStorageOptions<TValue>,
): UseLocalStorageResult<TValue> {
  // eslint-disable-next-line no-console
  console.error('`useLocalStorage` is not available.');

  return [initialValue, noop, noop];
}

export const useLocalStorage = canUseDOM() ? useLocalStorageHook : useLocalStorageSSR;
