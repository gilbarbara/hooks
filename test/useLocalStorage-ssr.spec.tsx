/**
 * @vitest-environment node
 */

import * as React from 'react';
import { renderToString } from 'react-dom/server';

import { useLocalStorage } from '../src/useLocalStorage';

function Component() {
  const [value] = useLocalStorage('root', 'foo');

  return <div>{value}</div>;
}

describe('useLocalStorage-ssr', () => {
  it('should return the initial value', () => {
    const view = renderToString(<Component />);

    expect(view).toMatchSnapshot();
  });
});
