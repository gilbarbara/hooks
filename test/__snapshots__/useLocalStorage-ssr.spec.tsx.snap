// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useLocalStorage-ssr > should return the initial value 1`] = `"<div>foo</div>"`;
